File: tictactoe.js
var tictactoe = {
	EMPTY: "E",
	X: "X",
	O: "O",

	SIZE: 3,

	O_TURN: 0,
	X_TURN: 1,

	get_piece: function(turn) {
		return (turn === tictactoe.O_TURN ?  tictactoe.O : tictactoe.X);
	}
}

var State = function(prev_state) {

	this.turn = tictactoe.X_TURN;

	/* 3x3 board as a 2D array */
	this.board = [[],[],[]];
	
	/* Initialise the board with empty */
	for (var i = 0; i < tictactoe.SIZE; i++) {
		for (var j = 0; j < tictactoe.SIZE; j++) {
			this.board[i][j] = tictactoe.EMPTY;
		}
	}

	this.isGameOver = function() {
		return (this.isRowOver() || this.isColOver() || this.isDiagOver());
	}

	this.isBoardFull = function() {
		for (var i = 0, var j = 0; i < tictactoe.SIZE; i++, j++) {
			if (this.board[i][j] === tictactoe.EMPTY) {
				return false;
			}
		}
		return true;
	}

	/* Horizontal game over */
	this.isRowOver = function() {
		for (var i = 0; i < tictactoe.SIZE; i++) {
			if (this.board[i][0] != tictactoe.EMPTY && this.board[i][0] === this.board[i][1] && this.board[i][1] == this.board[i][2]) {
				return true;
			}
		}
		return false;
	}

	/* Vertical game over */
	this.isColOver = function() {
		for (var i = 0; i < tictactoe.SIZE; i++) {
			if (this.board[0][i] != tictactoe.EMPTY && this.board[0][i] === this.board[1][i] && this.board[1][i] === this.board[2][i]) {
				return true;
			}
		}
		return false;
	}
	
	/* Diagonal game over */
	this.isDiagOver = function() {
		if (this.board[0][0] != tictactoe.EMPTY && this.board[0][0] === this.board[1][1] && this.board[1][1] === this.board[2][2]) {
			return true;
		}

		if (this.board[0][2] != tictactoe.EMPTY && this.board[0][2] === this.board[1][1] && this.board[1][1] === this.board[2][0]) {
			return true;
		}
		return false;
	}

	this.draw = function(ctx, width, height) {
		ui.drawGrid(ctx, width, height);

		for (var i = 0; i < tictactoe.SIZE; i++) {
			for (var j = 0; j < tictactoe.SIZE; j++) {
				var piece = this.board[i][j];
				var x = (1 + i*2)*width/6;
				var y = (1 + j*2)*height/6;
				if (piece === tictactoe.X) {
					ui.drawX(ctx, x, y);
				}
				else if (piece === tictactoe.O) {
					ui.drawO(ctx, x, y);
				}	
			}
		}
	}
}



var game = {

	state: undefined,

	init: function() {
		game.state = new State();

		/* Set up HTML canvas with mouseclick event */
		var canvas = document.getElementById("canvas");
		canvas.addEventListener("mousedown", function(event) {
			var x = event.x - canvas.offsetLeft;
			var y = event.y - canvas.offsetTop;
			
			x = Math.floor(x/(canvas.width/3));
			y = Math.floor(y/(canvas.height/3));

			game.make_move(x, y);
	
		}, false);
	},

	make_move: function(row, col) {
		var state = game.state;

		if (state.board[row][col] === tictactoe.EMPTY) {
			state.board[row][col] = tictactoe.get_piece(state.turn);
			game.draw();

			/* Game over check */
			if (game.state.isGameOver()) {
				alert(game.state.turn + " wins!");
				game.play();
				return;
			}
			else if (game.state.isBoardFull()) {
				alert("It's a draw!");
				game.play();
				return();
			}
			state.turn ^= 1;
		}
	},

	draw: function() {
		var canvas = document.getElementById("canvas");
		var ctx = canvas.getContext("2d");

		ctx.clearRect(0,0,canvas.width, canvas.height);
		game.state.draw(ctx, canvas.width, canvas.height);
	},

	play: function() {
		game.init();
		game.draw();
	}

}

File: tictactoe.py
class Tictactoe(object):

    def __init__(self):
        self.EMPTY = '_'
        self.X = 'X'
        self.O = 'O'

        self.SIZE = 3

        self.reset()
        

    def make_move(self, pos):
        if (pos < self.SIZE**2 and self.board[pos] == self.EMPTY):
            self.board[pos] = self.turn
            self.print_board()

            if self.is_gameover():
                print(self.turn + " won!")
            elif self.is_board_full():
                print("It's a draw!")
            else:
                self.turn = self.next_turn()
        else:
            print("Invalid move")

    def next_turn(self):
       return (self.O if self.turn == self.X else self.X)

    def is_gameover(self):
        return self.is_row_over() or self.is_col_over() or self.is_diag_over()

    def is_board_full(self):
        non_empty = [not self.board[i] == self.EMPTY for i in range(self.SIZE**2)]

        if all(non_empty):
            return True
        return False
    
    def reset(self): 
        self.turn = self.X
        self.board = [self.EMPTY for i in range(self.SIZE**2)]

        self.print_board()
      

    def is_row_over(self):
        for i in range(0, (self.SIZE**2), self.SIZE):
            same = [self.board[j] == self.board[j+1] for j in range(i, (i + self.SIZE -1))]
            
            if all(same) and not self.board[i] == self.EMPTY:
                return True
        return False

    def is_col_over(self):
        for i in range(self.SIZE):
            same = [self.board[j] == self.board[j+self.SIZE] for j in range(i, (self.SIZE**2 - self.SIZE), self.SIZE)]

            if all(same) and not self.board[i] == self.EMPTY:
                return True

        return False

    def is_diag_over(self):
        increment = self.SIZE - 1
        same = [self.board[i] == self.board[i+increment] for i in range(increment, (self.SIZE**2 - 2*increment), increment)]

        if all(same) and not self.board[increment] == self.EMPTY:
            return True

        increment = self.SIZE + 1
        same = [self.board[i] == self.board[i+increment] for i in range(0, (self.SIZE**2 - increment), increment)]

        if all(same) and not self.board[increment] == self.EMPTY:
            return True

        return False

    def print_board(self):
        for i in range(self.SIZE**2):
            if i % self.SIZE == 0 and not i == 0:
                print()
                print(''.join(['-' for i in range(self.SIZE*2 - 1)]))
            if not i % self.SIZE == 0:
                print('|', end='')
            print(self.board[i], end='')
        print()

